### PR TITLE
brickstore: init at 0-unstable-2024-05-02

### PR DESCRIPTION
**Pull request by jrobinson-uk**
## Description of changes

Created a new package deviation for building the popular offline Lego store inventory tool Brickstore.
First version begins at v2024.5.2, see [release notes](https://github.com/rgriebl/brickstore/releases/tag/v2024.5.2)

## Things done

Built on local system and tested through last 2 package releases

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [X] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

I would like to support by reviewing pull requests but would like to see this pull request through to ensure I understand the process
---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by DontEatOreo** [comment]
Thanks for making a PR! You will need to rebase so you can comply with the [Commit Conventions](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions)

In your case please squash d49b7d30f35aeabdb6bc0617a09a7bf29d50ecbe 7fed93b6e68f92ede655b9bb8319b9fdb8ac1376 9d16a5fb3c26655257d84b53b3bf0a05894878ae into bb280d1bc87504ad13b77d434abe62af3996023c and reword bb280d1bc87504ad13b77d434abe62af3996023c to `brickstore: init at 0-unstable-2024-05-02`

As well adding yourself as maintainer should be separate commit and in your case should be titled `maintainers: add legojames`

**Comment by DontEatOreo** [comment]
I'm not exactly sure how package this but I think this will set you in the right direction.

```nix
{
  lib,
  stdenv,
  qt6,
  libsForQt5,
  fetchFromGitHub,
  gst_all_1,
  cmake,
  libglvnd,
  tbb,
  ninja,
  pkg-config,
}:
let
  inherit (libsForQt5) qcoro;
in
stdenv.mkDerivation (finalAttrs: {
  pname = "brickstore";
  version = "0-unstable-2024-05-02";

  src = fetchFromGitHub {
    owner = "rgriebl";
    repo = finalAttrs.pname;
    rev = "v2024.5.2";
    hash = "sha256-Bu9oNbZm3lx/CfYAReHyWe/kW+kaefDWeBtWLHOCORU=";
    fetchSubmodules = true;
  };

  nativeBuildInputs = [
    cmake
    libglvnd
    ninja
    pkg-config
    qcoro
    qt6.qtdoc
    qt6.qtdeclarative
    qt6.qtimageformats
    qt6.qtmultimedia
    qt6.qtquick3d
    qt6.qtquicktimeline
    qt6.qtshadertools
    qt6.qttools
    qt6.qtwayland
    qt6.wrapQtAppsHook
    tbb
  ];

  preConfigure = ''
    # deletes line https://github.com/rgriebl/brickstore/blob/1e62a8bdf7da848a0033b8647d3711d95a3afd8d/cmake/BuildQCoro.cmake#L21
    # This isn't good but not sure how to do it via substituteInPlace
    sed -i '/^)$/d' cmake/BuildQCoro.cmake

    substituteInPlace cmake/BuildQCoro.cmake \
      --replace-fail 'FetchContent_Declare(' ' ' \
      --replace-fail '    qcoro' ' ' \
      --replace-fail '    GIT_REPOSITORY https://github.com/danvratil/qcoro.git' ' ' \
      --replace-fail '    GIT_TAG        v''${QCORO_VERSION}' ' ' \
      --replace-fail 'FetchContent_GetProperties(qcoro)' ' ' \
      --replace-fail 'FetchContent_Populate(qcoro)' ' ' \
      --replace-fail \
        'add_subdirectory(''${qcoro_SOURCE_DIR} ''${qcoro_BINARY_DIR} EXCLUDE_FROM_ALL)' \
        'add_subdirectory(${qcoro.src} ${qcoro}bin/qcoro)'
  '';

  qtWrapperArgs = [
    ''
      --prefix GST_PLUGIN_SYSTEM_PATH_1_0 : ${
        lib.makeLibraryPath [
          gst_all_1.gstreamer
          gst_all_1.gst-plugins-base
          gst_all_1.gst-plugins-good
          gst_all_1.gst-plugins-bad
          gst_all_1.gst-plugins-ugly
          gst_all_1.gst-libav
        ]
      }
    ''
  ];

  meta = {
    changelog = "https://github.com/rgriebl/brickstore/blob/main/CHANGELOG.md";
    description = "BrickLink offline management tool";
    homepage = "https://www.brickstore.dev/";
    license = lib.licenses.gpl3Plus;
    longDescription = ''
      BrickStore is a BrickLink offline management tool.
      It is multi-platform (Windows, macOS and Linux as well as iOS and Android),
      multilingual (currently English, German, Spanish, Swedish and French), fast and stable.
    '';
    maintainers = with lib.maintainers; [ legojames ];
    platforms = lib.platforms.linux;
  };
})
```

**Comment by jrobinson-uk** [comment]
> I'm not exactly sure how package this but I think this will set you in the right direction.
> 
> ```nix
> {
>   lib,
>   stdenv,
>   qt6,
>   libsForQt5,
>   fetchFromGitHub,
>   gst_all_1,
>   cmake,
>   libglvnd,
>   tbb,
>   ninja,
>   pkg-config,
> }:
> let
>   inherit (libsForQt5) qcoro;
> in
> stdenv.mkDerivation (finalAttrs: {
>   pname = "brickstore";
>   version = "0-unstable-2024-05-02";
> 
>   src = fetchFromGitHub {
>     owner = "rgriebl";
>     repo = finalAttrs.pname;
>     rev = "v2024.5.2";
>     hash = "sha256-Bu9oNbZm3lx/CfYAReHyWe/kW+kaefDWeBtWLHOCORU=";
>     fetchSubmodules = true;
>   };
> 
>   nativeBuildInputs = [
>     cmake
>     libglvnd
>     ninja
>     pkg-config
>     qcoro
>     qt6.qtdoc
>     qt6.qtdeclarative
>     qt6.qtimageformats
>     qt6.qtmultimedia
>     qt6.qtquick3d
>     qt6.qtquicktimeline
>     qt6.qtshadertools
>     qt6.qttools
>     qt6.qtwayland
>     qt6.wrapQtAppsHook
>     tbb
>   ];
> 
>   preConfigure = ''
>     # deletes line https://github.com/rgriebl/brickstore/blob/1e62a8bdf7da848a0033b8647d3711d95a3afd8d/cmake/BuildQCoro.cmake#L21
>     # This isn't good but not sure how to do it via substituteInPlace
>     sed -i '/^)$/d' cmake/BuildQCoro.cmake
> 
>     substituteInPlace cmake/BuildQCoro.cmake \
>       --replace-fail 'FetchContent_Declare(' ' ' \
>       --replace-fail '    qcoro' ' ' \
>       --replace-fail '    GIT_REPOSITORY https://github.com/danvratil/qcoro.git' ' ' \
>       --replace-fail '    GIT_TAG        v''${QCORO_VERSION}' ' ' \
>       --replace-fail 'FetchContent_GetProperties(qcoro)' ' ' \
>       --replace-fail 'FetchContent_Populate(qcoro)' ' ' \
>       --replace-fail \
>         'add_subdirectory(''${qcoro_SOURCE_DIR} ''${qcoro_BINARY_DIR} EXCLUDE_FROM_ALL)' \
>         'add_subdirectory(${qcoro.src} ${qcoro}bin/qcoro)'
>   '';
> 
>   qtWrapperArgs = [
>     ''
>       --prefix GST_PLUGIN_SYSTEM_PATH_1_0 : ${
>         lib.makeLibraryPath [
>           gst_all_1.gstreamer
>           gst_all_1.gst-plugins-base
>           gst_all_1.gst-plugins-good
>           gst_all_1.gst-plugins-bad
>           gst_all_1.gst-plugins-ugly
>           gst_all_1.gst-libav
>         ]
>       }
>     ''
>   ];
> 
>   meta = {
>     changelog = "https://github.com/rgriebl/brickstore/blob/main/CHANGELOG.md";
>     description = "BrickLink offline management tool";
>     homepage = "https://www.brickstore.dev/";
>     license = lib.licenses.gpl3Plus;
>     longDescription = ''
>       BrickStore is a BrickLink offline management tool.
>       It is multi-platform (Windows, macOS and Linux as well as iOS and Android),
>       multilingual (currently English, German, Spanish, Swedish and French), fast and stable.
>     '';
>     maintainers = with lib.maintainers; [ legojames ];
>     platforms = lib.platforms.linux;
>   };
> })
> ```

Is this to replace my patch approach? I took that approach as I thought I read something that said that downloading addtional requirements should be done using nixos helper tools like fetchFromGithub

**Comment by DontEatOreo** [comment]
Your older commits had a patch file, but it was just a duplicate of the `default.nix` file with the `.patch` extension (As seen in d49b7d30f35aeabdb6bc0617a09a7bf29d50ecbe). There was no actual patch in that commit, so I don't know what your original approach was. After looking at the project, I found that only `cmake/BuildQCoro.cmake` was trying to download something from GitHub and I just got rid of that logic and patch the path for `qcoro`

I noticed you were downloading `qcoro` with `fetchFromGitHub` to patch the source code. This isn't needed as `qcoro` is already packaged in nixpkgs

I was able to build the package on `x86_64-linux` and run the GUI program, but I don't know how to use it. So I don't know if everything is working correctly. Please let me know if there are any issues.

**Comment by jrobinson-uk** [comment]
> brickstore: init at 0-unstable-2024-05-02

I've never had to rebase before, I've tried `git rebase -i master` and squashed the relevant commits but it leads to merge conflicts from other parts of nix-pkgs.

Would like to figure it out but would it be easier to close this pull, re-clone my fork and start again?

**Comment by jrobinson-uk** [comment]
> git rebase -i HEAD~5

I get `fatal: invalid upstream 'HEAD-5' `

**Comment by DontEatOreo** [comment]
That's because you typed  `-` instead of `~`

**Comment by jrobinson-uk** [comment]
> brickstore: init at 0-unstable-2024-05-02

I'm an idiot.

Think I've done all that and now just need to push changes, which in turn requires a pull. Do I need to merge, rebase or fast-forward.

```
hint: You have divergent branches and need to specify how to reconcile them.
hint: You can do so by running one of the following commands sometime before
hint: your next pull:
hint: 
hint:   git config pull.rebase false  # merge
hint:   git config pull.rebase true   # rebase
hint:   git config pull.ff only       # fast-forward only
```

**Comment by DontEatOreo** [comment]
No worries! Run `git reflog`. This will open up a log of the git commands you've done. Afterwards, you just need to find which line was the one before the merge conflict and then copy the "HEAD." It will look something like this: `HEAD@{X}`. After that, you will just need to run `git reset --hard HEAD@{X}`, which allows you to essentially "go back in time" and redo the rebase again, and then you can force push the changes.

**Comment by DontEatOreo** [comment]
I see you're struggling a bit, but no worries! So make sure to run `git reflog` and take ***a close look*** at which line was the one before all the extra commits, **the one where you had your original 5 commits**, then just do `git reset --hard HEAD@{X}` *(`X` being whatever number you need based on your local history)* and start the rebase again, *you don't need to make any extra commits*.

**Comment by jrobinson-uk** [comment]
> git reflog

Thanks, I think that's HEAD{742} based on this:

```
d49b7d30f35a HEAD@{741}: commit: brickstore: v2024.5.2
7fed93b6e68f HEAD@{742}: commit: brickstore: init at 2024.5.2
```

**Comment by DontEatOreo** [comment]
It's a bit hard for me to tell exactly, but it seems that `HEAD@{742}` is the one. After resetting to it, take a look at your `git log` and squash any extra commits into your initial commit. Then "softly reset" the commit, then commit the maintainer file, and then your package, after that ***make sure to force push your changes***.

**Comment by jrobinson-uk** [comment]
> brickstore: init at 0-unstable-2024-05-02

I think I've done that now :crossed_fingers: , thanks for the help in getting this far

**Comment by DontEatOreo** [comment]
Almost correct! Reset again *but carefully rebase,* as to not pick up anyone else commits

**Comment by jrobinson-uk** [comment]
> Almost correct! Reset again _but carefully rebase,_ as to not pick up anyone else commits

Is that just down to the number of commits in the `HEAD~5` bit I thought is was just viewing the last 5, but is it actually "include last 5 commits"?

**Comment by DontEatOreo** [comment]
When I was originally asking you to run that command, there were 5 commits, hence the `HEAD~5`. The numbering starts from 1, where 1 is the latest commit *(as `HEAD~1`)*, and as you increment the number, you go further down the commit history

**Comment by DontEatOreo** [comment]
You just need to squash 9d16a5fb3c26655257d84b53b3bf0a05894878ae

**Comment by jrobinson-uk** [comment]
> You just need to squash [9d16a5f](https://github.com/NixOS/nixpkgs/commit/9d16a5fb3c26655257d84b53b3bf0a05894878ae)

I was sure I did, hmmm

So...
```
git reset --hard HEAD@{812}
Updating files: 100% (45716/45716), done.
HEAD is now at bb280d1bc875 brickstore: v2024.5.2 Fixing editorconfig issue
```

then `git rebase -i HEAD~5` produces:
```
pick 229bf85d5fa8 rustpython: 0.3.0 -> 0.3.1
pick 9d16a5fb3c26 Last Sync: 2024-05-07 21:00:42 (Xen)
pick 7fed93b6e68f brickstore: init at 2024.5.2 https://github.com/rgriebl/brickstore/releases/tag/v2024.5.2
pick d49b7d30f35a brickstore: v2024.5.2 Adding no-download patch to ensure qcoro is fetched and built via fetchFromGitHub
pick bb280d1bc875 brickstore: v2024.5.2 Fixing editorconfig issue
```

which goes to far, so then I `git rebase --abort` and `git rebase -i HEAD~4`, which I edit to:
```
pick 9d16a5fb3c26 Last Sync: 2024-05-07 21:00:42 (Xen)
squash 7fed93b6e68f brickstore: init at 2024.5.2 https://github.com/rgriebl/brickstore/releases/tag/v2024.5.2
squash d49b7d30f35a brickstore: v2024.5.2 Adding no-download patch to ensure qcoro is fetched and built via fetchFromGitHub
squash bb280d1bc875 brickstore: v2024.5.2 Fixing editorconfig issue
```
then 
```
git reset --soft HEAD~1
git rm pkgs/applications/misc/brickstore/default.nix  -f
git rm pkgs/applications/misc/brickstore/no-download.patch -f
git commit -m "maintainers: add legojames!
mkdir pkgs/by-name/br/brickstore
cp /etc/nixos/pkgs/brickstore.nix pkgs/by-name/br/brickstore/package.nix
git add pkgs/by-name/br/brickstore/package.nix
git commit  -m "brickstore: init at 0-unstable-2024-05-02"
git push --force
```

**Comment by DontEatOreo** [comment]
Seems like you were able to solve the last issue without me. I'm happy to say now that the commits are perfect. Great job!

File: pkgs/by-name/br/brickstore/package.nix
Comment (DontEatOreo) [inline pkgs/by-name/br/brickstore/package.nix:59]:
```suggestion
  preConfigure = ''
    # deletes line https://github.com/rgriebl/brickstore/blob/1e62a8bdf7da848a0033b8647d3711d95a3afd8d/cmake/BuildQCoro.cmake#L21
    # This isn't good but not sure how to do it via substituteInPlace
    sed -i '/^)$/d' cmake/BuildQCoro.cmake
```

Make a new commit with the changes and squash it using `git rebase -i HEAD~2`.

**Comment by DontEatOreo** [comment]
Please update your PR title to be `brickstore: init at 0-unstable-2024-05-02`

**Comment by jrobinson-uk** [comment]
> https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions

Have I missed something?

**Comment by DontEatOreo** [comment]
> If you have commits pkg-name: oh, forgot to insert whitespace: squash commits in this case. Use git rebase -i. See [Squashing Commits](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History#_squashing) for additional information.

Please rebase and squash the other commits into e28daba898e728c78aebc2791e388f813b60f967

**Comment by DontEatOreo** [comment]
Result of `nixpkgs-review pr 309937` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>brickstore</li>
  </ul>
</details>

**Review by 4evy — APPROVED** [review]

**Comment by DontEatOreo** [comment]
LGTM, Thank you for sticking with the PR, even though I know it was a bit challenging!

**Comment by jrobinson-uk** [comment]
> LGTM, Thank you for sticking with the PR, even though I know it was a bit challenging!

Yay, thanks you so much for the support, I learnt a lot (slowly) and feel more confident about adding packages going forward.

What happens next with this PR, how does this become installable from nixpkgs for other users?

**Comment by DontEatOreo** [comment]
Somebody with commit access has to merge it, so there are a couple of scenarios:

- You just wait until someone with commit access finds this PR and decides to merge it.
- You can post the PR on https://discourse.nixos.org/t/prs-already-reviewed/2617 and again wait for some with commit access to merge it.

For future PR's that haven't been reviewed yet, you can ask on https://discourse.nixos.org/t/prs-ready-for-review/3032 or if you have Matrix account you can ask here too: https://app.element.io/#/room/#review-requests:nixos.org [^1]

I added a backport label, which will automatically create a backport PR that will "back port" the program to nixpkgs unstable. You can track the progress here: https://nixpk.gs/pr-tracker.html?pr=309937

When it gets merged into `master`, you can skip waiting for the backport to be merged, by temporarily installing the package with `nix profile install github:NixOS/nixpkgs/master#brickstore`[^2] and after it arrives in the channel (or inputs if you use flakes), you can remove it from the profile.

And final note to save time and effort for both yourself and reviewers, please ***take a careful look again*** at [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#commit-conventions) so there is less back and forth, and future PR can get merged faster!

### Wishing you luck for any future PRs!

[^1]: If it takes too long for it to be merged, you can also post in the review requests channels as well.
[^2]: You will need to have the new `nix3` commands enabled for this, though:
https://nixos.org/manual/nix/stable/contributing/experimental-features.html?highlight=experimental-features
https://nixos.org/manual/nix/stable/command-ref/new-cli/nix3-profile

File: pkgs/by-name/br/brickstore/package.nix
Comment (drupol) [inline pkgs/by-name/br/brickstore/package.nix:80]: Missing `meta.mainProgram`

File: pkgs/by-name/br/brickstore/package.nix
Comment (drupol) [inline pkgs/by-name/br/brickstore/package.nix:23]: Please don't use this.

`repo = pname;` is nice for DRY but creates a binding that goes too far.

See further information about this here: https://github.com/nix-community/nixpkgs-lint/issues/21

**Review by drupol — CHANGES_REQUESTED** [review]
Hi,

Thanks for your PR. I've left some comments, let me know if this is clear enough.

**Review by drupol — APPROVED** [review]

**Comment by drupol** [comment]
Please rebase your branch on top of the latest commit from `master` and it's good for me.

**Comment by drupol** [comment]
Nope... this PR shouldn't contain any merge commit.

**Review by 4evy — APPROVED** [review]
LGTM!

In the future, when you have merge conflict on a PR, you can checkout `master` then pull with a rebase (make sure to update your fork master first) and then do a `git rebase` on your local branch.

```console
git checkout master
# you can add `origin master` to be very explicit
git pull --rebase
git checkout add-bricktore-package
git rebase master
```

Then resolve any conflict

File: pkgs/by-name/br/brickstore/package.nix
Comment (drupol) [inline pkgs/by-name/br/brickstore/package.nix:58]: Please fix this as mentionned in the previous comment.

**Review by drupol — CHANGES_REQUESTED** [review]

**Review by drupol — APPROVED** [review]

**Pull request merged**
